### PR TITLE
fix api versions for some robowebsettings methods

### PR DESCRIPTION
**Pull request by bdwain**
some robowebsettings.java.vm methods were incorrectly restricted to higher api levels. this fixes them. see #2097

### CODE REVIEW

**Comment by bdwain** [comment]
looks like this is passing. though i was a little confused when i looked at the file that I edited. a bunch of the methods in the section for API level 22 did not exist in the [android documentation for WebSettings](http://developer.android.com/reference/android/webkit/WebSettings.html), yet they were marked as @Override. (e.g. setUserAgent. This method does not exist on WebSettings). They are all in a #if $api >= 22 for now. It seems like the build must just not be building for API 22, and that's why it's passing. I left them in, but I'm not really sure why they are there.

**Comment by 1zaman** [comment]
> a bunch of the methods in the section for API level 22 did not exist in the [android documentation for WebSettings](http://developer.android.com/reference/android/webkit/WebSettings.html), yet they were marked as Override. (e.g. setUserAgent. This method does not exist on WebSettings). They are all in a #if $api >= 22 for now.

As I mentioned in the issue, these methods were deprecated and hidden from the public API documentation. However, hidden methods are still visible to Robolectric, so overriding them does not result in a compiler error, and we can still let them remain I guess (just like all the other implementations of hidden methods).

> It seems like the build must just not be building for API 22, and that's why it's passing. I left them in, but I'm not really sure why they are there.

Actually the error was caused by your removing the API definition from `setAcceptThirdPartyCookies()` and `getAcceptThirdPartyCookies()` which are hidden system API methods that were introduced in Lollipop, and thus need to be restricted to build against API 21+ only.

**Comment by bdwain** [comment]
so the only change you're suggesting is to setAcceptThirdPartyCookies() and getAcceptThirdPartyCookies() in a block for API 21+?

**Comment by 1zaman** [comment]
@bdwain: You can also remove the API 22 restriction from `getUserAgent()`, `setUserAgent()`, `getNavDump()`, `setNavDump()`, `getUseWebViewBackgroundForOverscrollBackground()`, and `setUseWebViewBackgroundForOverscrollBackground`()` methods. Their initial API level is much lower; the first four were present from API 1, while the last was introduced in Gingerbread.

**Comment by 1zaman** [comment]
Otherwise good job on this pull request. Looks like it addresses the issue of wrong API level definitions. However, there are lots of methods that were introduced at a specific API level, that are not defined as such in `RoboWebSettings`. This should also be addressed, preferably in this PR.

**Comment by bdwain** [comment]
@1zaman if you look at my first commit, [it failed](https://travis-ci.org/robolectric/robolectric/jobs/86514379). if removing the api 22 checks is ok, do you think it was just from the thirdpartycookies methods?

**Comment by 1zaman** [comment]
> do you think it was just from the thirdpartycookies methods?

Yes, I think so.

**Comment by bdwain** [comment]
cool. thanks. I'll work on getting getting each method in the right if $api block

**Comment by bdwain** [comment]
organized the file into blocks for each api level that added any new functionality

File: robolectric-shadows/shadows-core/src/main/resources/org/robolectric/fakes/RoboWebSettings.java.vm [hunk 1]
```diff
@@ -602,24 +516,117 @@ public class RoboWebSettings extends WebSettings {
...
+  }
+
+  @Override
+  public int getMixedContentMode() {
+    return mixedContentMode;
   }
 
-  // End API 22
+  @Override
+  public void setVideoOverlayForEmbeddedEncryptedVideoEnabled(boolean b) {
```
Comment (1zaman) [inline robolectric-shadows/shadows-core/src/main/resources/org/robolectric/fakes/RoboWebSettings.java.vm]: I guess this should ideally change an actual flag that is returned in `getVideoOverlayForEmbeddedEncryptedVideoEnabled()`.

**Comment by 1zaman** [comment]
:+1: Thanks for all the hard work on fixing this!

File: robolectric-shadows/shadows-core/src/main/resources/org/robolectric/fakes/RoboWebSettings.java.vm (hunk 1 above)
Comment (bdwain) [inline robolectric-shadows/shadows-core/src/main/resources/org/robolectric/fakes/RoboWebSettings.java.vm]: i wasn't sure if that was hard coded to false on purpose. it was like that before. i can switch it though

**Comment by bdwain** [comment]
i think this should be good to go. everything is restricted to its proper api now.

**Comment by bdwain** [comment]
can this go out as a patch to 3.0?

**Comment by erd** [comment]
No

**Comment by bdwain** [comment]
Well do you know what the eta on 3.1 would be? You can't really use robolectric with a websettings object right now

**Comment by erd** [comment]
API 23 support is the main thing that's driving 3.1. I don't have an eta on it.

**Comment by bdwain** [comment]
Any recommendations in the mean time?

**Comment by erd** [comment]
You should be able to build master (that includes these changes - or take them from the latest Travis build), check them in locally and set the property `robolectric.dependency.dir` so Robolectric fetches from that directory instead of from Maven Central. There are some (albeit brief) docs at http://robolectric.org/configuring.

**Comment by bdwain** [comment]
Cool I'll try that tomorrow. Thanks!

**Comment by 1zaman** [comment]
@erd: Shouldn't it be available from the SNAPSHOT artifacts on Sonatype OSS?

**Comment by erd** [comment]
Yes, that would also work.

**Comment by bdwain** [comment]
are there instructions for accessing this anywhere?

**Comment by 1zaman** [comment]
@bdwain: You should be able to add the SNAPSHOT artifact as follows:

1. Add https://oss.sonatype.org/content/repositories/snapshots/ to your repositories list.
2. Define `3.1-SNAPSHOT` as the version for all your Robolectric modules.

**Comment by bdwain** [comment]
thanks ill try that.

**Pull request closed without merging**